src/datasources/erc721.ts

```diff
@@ -88,87 +94,89 @@ export function handleConsecutiveTransfer(event: ConsecutiveTransfer): void {
 	let contract = fetchERC721(event.address)
-	if (contract != null) {
-		let owner      = fetchAccount(event.params.owner)
-		let operator   = fetchAccount(event.params.operator)
-		let delegation = fetchERC721Operator(contract, owner, operator)
-
-		delegation.approved = event.params.approved
-
-		delegation.save()
-
-		// 	let ev = new ApprovalForAll(events.id(event))
-		// 	ev.emitter     = contract.id
-		// 	ev.transaction = transactions.log(event).id
-		// 	ev.timestamp   = event.block.timestamp
-		// 	ev.delegation  = delegation.id
-		// 	ev.owner       = owner.id
-		// 	ev.operator    = operator.id
-		// 	ev.approved    = event.params.approved
-		// 	ev.save()
-	}
+	if (contract == null) return
+
+	let owner      = fetchAccount(event.params.owner)
+	let operator   = fetchAccount(event.params.operator)
+	let delegation = fetchERC721Operator(contract, owner, operator)
+
+	delegation.approved = event.params.approved
+
+	delegation.save()
+
+	// 	let ev = new ApprovalForAll(events.id(event))
+	// 	ev.emitter     = contract.id
+	// 	ev.transaction = transactions.log(event).id
+	// 	ev.timestamp   = event.block.timestamp
+	// 	ev.delegation  = delegation.id
+	// 	ev.owner       = owner.id
+	// 	ev.operator    = operator.id
+	// 	ev.approved    = event.params.approved
+	// 	ev.save()
 }
 
 export function handleMetadataUpdate(event: MetadataUpdateEvent) : void {
-	let erc721   = IERC721.bind(event.address)
 	let contract = fetchERC721(event.address)
+	if (contract == null) return
 
-	if (contract != null) {
-		if (contract.supportsMetadata) {
-			let token        = fetchERC721Token(contract, event.params._tokenId)
-			let try_tokenURI = erc721.try_tokenURI(event.params._tokenId)
-			token.uri        = try_tokenURI.reverted ? '' : try_tokenURI.value
-			token.save()
-		} else {
-			// add a warning ?
-		}
+	if (contract.supportsMetadata) {
+		_updateURI(contract, event.params._tokenId)
+	} else {
+		// add a warning ?
 	}
 }
 
 export function handleBatchMetadataUpdate(event: BatchMetadataUpdateEvent) : void {
-	let erc721   = IERC721.bind(event.address)
 	let contract = fetchERC721(event.address)
-
-	if (contract != null) {
-		if (contract.supportsMetadata) {
-			const from = event.params._fromTokenId.toU64();
-			const to   = event.params._toTokenId.toU64();
-			// Updates of blocks larger than 5000 tokens may DoS the subgraph, we skip them
-			if (to - from <= 5000) {
-				for (let tokenId =from; tokenId <= to; ++tokenId) {
-					let token        = fetchERC721Token(contract, BigInt.fromU64(tokenId))
-					let try_tokenURI = erc721.try_tokenURI(BigInt.fromU64(tokenId))
-					token.uri        = try_tokenURI.reverted ? '' : try_tokenURI.value
-					token.save()
-				}
+	if (contract == null) return
+
+	if (contract.supportsMetadata) {
+		let fromTokenId = event.params._fromTokenId.toU64()
+		let toTokenId   = event.params._toTokenId.toU64()
+		// Updates of blocks larger than 5000 tokens may DoS the subgraph, we skip them
```

Comment:
We've discussed this value (5000) previously since this is the reason for the ERC721Consecutive cap, but I never knew how we got to it. Would you mind clarifying?

Reply:
My concern is that if someone create an event with `fromTokenId = 0`, and `toTokenId = maxU64` ... then the for loop would iterate for 2**64 loops.

Considering this loop does a db lookup for the token, and then does an update, with a write to the database ... I'm not sure it is realistic to do that much more than 1000 times per second

Lets assume we still manage to update 1 million tokens seconds. At that rate, updating 2**64 tokens would take 584942 years !!! This would cause the subgraph to halt.

The fact that just one simple event, that anyone could emit, is able to take down a subgraph is a real issue.

The protection is obvious: reject updates that are too large. I set an arbitrary limit of 5000 tokens that matches what we observed on other indexing platform, and that we included in our implementation of batchMinting.